sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Canonicalize.scala

```diff
@@ -46,6 +47,15 @@ object Canonicalize {
     case _ => e
   }
 
+  /** Remove TimeZoneId for Cast if needsTimeZone return false. */
+  private[expressions] def ignoreTimeZone(e: Expression): Expression = e match {
+    case a: AnsiCast if !a.needsTimeZone && a.timeZoneId.nonEmpty =>
+      a.copy(timeZoneId = None)
+    case c: Cast if !c.needsTimeZone && c.timeZoneId.nonEmpty =>
+      c.copy(timeZoneId = None)
```

Comment:
```scala
case c: CastBase if c.timeZoneId.nonEmpty && !c.needsTimeZone =>
      c.copy(timeZoneId = None)
```
No need to call `needsTimeZone` twice.

Reply:
`CastBase` is not a case class and can't call `copy`

Reply:
hmm, ok:
```scala
    case c: CastBase if c.timeZoneId.nonEmpty && !c.needsTimeZone => c.withTimeZone(null)
```

Reply:
Thanks, done in 73f4694.